src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1553,10 +1556,35 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
                 ->addTag('messenger.receiver', array('alias' => $name))
                 ->addTag('messenger.sender', array('alias' => $name))
             ;
+
+            if ('symfony://kernel.terminate' === $transport['dsn']) {
+                $busId = $transport['options']['bus'] ?? null;
+                if (null === $busId) {
+                    throw new InvalidArgumentException(sprintf('Missing mandatory "bus" option for transport "%s"', $name));
+                }
+                if (!$container->hasDefinition($busId)) {
+                    throw new InvalidArgumentException(sprintf('No bus was found with id: %s', $busId));
+                }
+
+                $transportDefinition->addTag('kernel.event_listener', array('event' => KernelEvents::TERMINATE, 'method' => 'flush'));
+                $transportDefinition->addTag('kernel.event_listener', array('event' => KernelEvents::EXCEPTION, 'method' => 'stop'));
+                $shouldConfigureKernelTerminateFactory = true;
+            }
+
             $container->setDefinition($transportId = 'messenger.transport.'.$name, $transportDefinition);
             $senderAliases[$name] = $transportId;
         }
 
+        if ($shouldConfigureKernelTerminateFactory) {
```

Comment:
This optimization looks unnecessary to me. You can always replace the argument right after buses being registered.

Reply:
Because it actually doesn't matter. :) 
Note you don't remove the `messenger.transport.kernel_terminate.factory` despite not being used if no kernel.terminate transport is configured. So it will exist and partially configured (missing the service locator).

As a general note, Symfony will either remove an unused definition, or inline it if used only once.
Here `messenger.transport.kernel_terminate.factory` will be inlined and removing it would not provide much benefit anyway.

Regarding the service locator itself, `ServiceLocatorTagPass::register` also optimize things under the hood, and will only generate one with the exact same arguments, shared between different services needing a bus locator, like the `ConsumeMessagesCommand` command for instance.